Add signing mechanism to livepatch-build-tools.

The process for adding the signature is:
- Add a new .note.<product-name>.signature section to the livepatch with an empty signature
- Calculate the signature of the livepatch with the new section using a private key
- Update the signature section to include the new signature

Signed-off-by: Gerald Elder-Vass <gerald.elder-vass@cloud.com>
diff --git a/.gitignore b/.gitignore
index e53b55e391a9..d2a6e11351ae 100644
--- a/.gitignore
+++ b/.gitignore
@@ -2,7 +2,6 @@ create-diff-object
 prelink
 *.o
 *.d
-sign/extract-cert
-sign/sign-file
 sign/signing_key.pem
-sign/signing_key.x509
+sign/signing_cert.pem
+sign/signature.dat
diff --git a/Makefile b/Makefile
index 2f74c3433336..c1bb33afb0f0 100644
--- a/Makefile
+++ b/Makefile
@@ -15,7 +15,7 @@ LDFLAGS = -lelf
 LDFLAGS += $(shell if test -z "`$(CC) -largp 2>&1 | grep '\-largp'`"; \
                       then echo "-largp"; fi;)
 
-TARGETS = create-diff-object prelink
+TARGETS = create-diff-object prelink sign/livepatch-sign
 CREATE_DIFF_OBJECT_OBJS = create-diff-object.o lookup.o insn/insn.o insn/inat.o common.o
 PRELINK_OBJS = prelink.o lookup.o insn/insn.o insn/inat.o common.o
 SOURCES = create-diff-object.c prelink.c lookup.c insn/insn.c insn/inat.c common.c
@@ -37,7 +37,7 @@ install: all
 	$(INSTALL) -d $(LIBEXECDIR)
 	$(INSTALL) $(TARGETS) livepatch-gcc $(LIBEXECDIR)
 	$(INSTALL) -d $(BINDIR)
-	$(INSTALL) livepatch-build $(BINDIR)
+	$(INSTALL) livepatch-build sign/livepatch-sign $(BINDIR)
 
 clean:
 	$(RM) $(TARGETS) $(CREATE_DIFF_OBJECT_OBJS) $(PRELINK_OBJS) *.d insn/*.d
diff --git a/sign/Makefile b/sign/Makefile
index d29761a84a48..c3b6174ea5cc 100644
--- a/sign/Makefile
+++ b/sign/Makefile
@@ -1,37 +1,3 @@
-SHELL = /bin/sh
-CC    = gcc
-
-.PHONY: all install clean distclean
-.DEFAULT: all
-
-CFLAGS  += -Iinsn -Wall -g
-LDFLAGS = -lcrypto
-
-CERTS = signing_key.pem signing_key.x509
-TARGETS = sign-file extract-cert $(CERTS)
-SIGN_OBJS = sign-file.o
-EXTRACT_CERT_OBJS = extract-cert.o
-SOURCES = sign-file.c extract-cert.c
-
-all: $(TARGETS)
-
--include $(SOURCES:.c=.d)
-
-%.o : %.c
-	$(CC) -MMD -MP $(CFLAGS) -c -o $@ $<
-
-sign-file: $(SIGN_OBJS)
-	$(CC) $(CFLAGS) $^ -o $@ $(LDFLAGS)
-
-extract-cert: $(EXTRACT_CERT_OBJS)
-	$(CC) $(CFLAGS) $^ -o $@ $(LDFLAGS)
-
-clean:
-	$(RM) $(TARGETS) $(SIGN_OBJS) $(EXTRACT_CERT_OBJS) *.d insn/*.d
-
-distclean: clean
-	$(RM) $(CERTS)
-
 # Generate a certificate if needed
 # These routines come from Linux.
 signing_key.pem: x509.genkey
@@ -45,7 +11,7 @@ signing_key.pem: x509.genkey
 	@echo "###"
 	openssl req -new -nodes -utf8 -sha256 -days 36500 \
 	        -batch -x509 -config x509.genkey \
-	        -outform PEM -out signing_key.pem \
+	        -outform PEM -out signing_cert.pem \
 	        -keyout signing_key.pem 2>&1
 	@echo "###"
 	@echo "### Key pair generated."
@@ -71,6 +37,3 @@ x509.genkey:
 	@echo >>$@ "subjectKeyIdentifier=hash"
 	@echo >>$@ "authorityKeyIdentifier=keyid"
 .INTERMEDIATE: x509.genkey
-
-signing_key.x509: extract-cert signing_key.pem
-	./extract-cert signing_key.pem $@
diff --git a/sign/extract-cert.c b/sign/extract-cert.c
deleted file mode 100644
index b071bf476fea..000000000000
--- a/sign/extract-cert.c
+++ /dev/null
@@ -1,162 +0,0 @@
-/* Extract X.509 certificate in DER form from PKCS#11 or PEM.
- *
- * Copyright © 2014-2015 Red Hat, Inc. All Rights Reserved.
- * Copyright © 2015      Intel Corporation.
- *
- * Authors: David Howells <dhowells@redhat.com>
- *          David Woodhouse <dwmw2@infradead.org>
- *
- * This program is free software; you can redistribute it and/or
- * modify it under the terms of the GNU Lesser General Public License
- * as published by the Free Software Foundation; either version 2.1
- * of the licence, or (at your option) any later version.
- */
-#define _GNU_SOURCE
-#include <stdio.h>
-#include <stdlib.h>
-#include <stdint.h>
-#include <stdbool.h>
-#include <string.h>
-#include <err.h>
-#include <openssl/bio.h>
-#include <openssl/pem.h>
-#include <openssl/err.h>
-#include <openssl/engine.h>
-
-#define PKEY_ID_PKCS7 2
-
-static __attribute__((noreturn))
-void format(void)
-{
-	fprintf(stderr,
-		"Usage: scripts/extract-cert <source> <dest>\n");
-	exit(2);
-}
-
-static void display_openssl_errors(int l)
-{
-	const char *file;
-	char buf[120];
-	int e, line;
-
-	if (ERR_peek_error() == 0)
-		return;
-	fprintf(stderr, "At main.c:%d:\n", l);
-
-	while ((e = ERR_get_error_line(&file, &line))) {
-		ERR_error_string(e, buf);
-		fprintf(stderr, "- SSL %s: %s:%d\n", buf, file, line);
-	}
-}
-
-static void drain_openssl_errors(void)
-{
-	const char *file;
-	int line;
-
-	if (ERR_peek_error() == 0)
-		return;
-	while (ERR_get_error_line(&file, &line)) {}
-}
-
-#define ERR(cond, fmt, ...)				\
-	do {						\
-		bool __cond = (cond);			\
-		display_openssl_errors(__LINE__);	\
-		if (__cond) {				\
-			err(1, fmt, ## __VA_ARGS__);	\
-		}					\
-	} while(0)
-
-static const char *key_pass;
-static BIO *wb;
-static char *cert_dst;
-int kbuild_verbose;
-
-static void write_cert(X509 *x509)
-{
-	char buf[200];
-
-	if (!wb) {
-		wb = BIO_new_file(cert_dst, "wb");
-		ERR(!wb, "%s", cert_dst);
-	}
-	X509_NAME_oneline(X509_get_subject_name(x509), buf, sizeof(buf));
-	ERR(!i2d_X509_bio(wb, x509), "%s", cert_dst);
-	if (kbuild_verbose)
-		fprintf(stderr, "Extracted cert: %s\n", buf);
-}
-
-int main(int argc, char **argv)
-{
-	char *cert_src;
-
-	OpenSSL_add_all_algorithms();
-	ERR_load_crypto_strings();
-	ERR_clear_error();
-
-	kbuild_verbose = atoi(getenv("KBUILD_VERBOSE")?:"0");
-
-        key_pass = getenv("KBUILD_SIGN_PIN");
-
-	if (argc != 3)
-		format();
-
-	cert_src = argv[1];
-	cert_dst = argv[2];
-
-	if (!cert_src[0]) {
-		/* Invoked with no input; create empty file */
-		FILE *f = fopen(cert_dst, "wb");
-		ERR(!f, "%s", cert_dst);
-		fclose(f);
-		exit(0);
-	} else if (!strncmp(cert_src, "pkcs11:", 7)) {
-		ENGINE *e;
-		struct {
-			const char *cert_id;
-			X509 *cert;
-		} parms;
-
-		parms.cert_id = cert_src;
-		parms.cert = NULL;
-
-		ENGINE_load_builtin_engines();
-		drain_openssl_errors();
-		e = ENGINE_by_id("pkcs11");
-		ERR(!e, "Load PKCS#11 ENGINE");
-		if (ENGINE_init(e))
-			drain_openssl_errors();
-		else
-			ERR(1, "ENGINE_init");
-		if (key_pass)
-			ERR(!ENGINE_ctrl_cmd_string(e, "PIN", key_pass, 0), "Set PKCS#11 PIN");
-		ENGINE_ctrl_cmd(e, "LOAD_CERT_CTRL", 0, &parms, NULL, 1);
-		ERR(!parms.cert, "Get X.509 from PKCS#11");
-		write_cert(parms.cert);
-	} else {
-		BIO *b;
-		X509 *x509;
-
-		b = BIO_new_file(cert_src, "rb");
-		ERR(!b, "%s", cert_src);
-
-		while (1) {
-			x509 = PEM_read_bio_X509(b, NULL, NULL, NULL);
-			if (wb && !x509) {
-				unsigned long err = ERR_peek_last_error();
-				if (ERR_GET_LIB(err) == ERR_LIB_PEM &&
-				    ERR_GET_REASON(err) == PEM_R_NO_START_LINE) {
-					ERR_clear_error();
-					break;
-				}
-			}
-			ERR(!x509, "%s", cert_src);
-			write_cert(x509);
-		}
-	}
-
-	BIO_free(wb);
-
-	return 0;
-}
diff --git a/sign/livepatch-sign b/sign/livepatch-sign
new file mode 100755
index 000000000000..d38cd4eaaa7d
--- /dev/null
+++ b/sign/livepatch-sign
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
+
+# Formatting of ELF note sections:
+# https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-18048.html
+
+# The note description field will be packed according to the following struct
+#
+# struct signature {
+#   uint16_t version;
+#   uint8_t sig_algo;
+#   uint8_t hash_algo;
+#   uint32_t length;
+#   uint8_t data[];
+# }
+
+import re
+import sys
+import struct
+import subprocess
+
+from enum import Enum
+
+
+if len(sys.argv) != 5:
+    print(f"Usage: {sys.argv[0]} <sig-algo> <hash-algo> <livepatch> <signature>", file=sys.stderr)
+    sys.exit(1)
+
+(sig_algo, hash_algo, livepatch, signature_binary_file) = sys.argv[1:]
+
+NAME = "Xen"
+SECTION_NAME = f".note.{NAME}.signature"
+SIGNATURE_FILE = "signature.dat"
+ALIGN = 4
+
+
+class HashAlgorithm(Enum):
+    sha256 = 0
+
+
+class SignatureAlgorithm(Enum):
+    rsa = 0
+
+
+def paddedLength(string):
+    return -ALIGN * (-len(string) // ALIGN)
+
+def assertPadded(string, message):
+    if len(string) % ALIGN != 0:
+        print(message, file=sys.stderr)
+        sys.exit(1)
+
+def packV1(sig_algo, hash_algo, signature):
+    contents = struct.pack("H", 1)  # Version (uint16_t)
+    contents += struct.pack("B", SignatureAlgorithm[sig_algo].value)  # SignatureAlgorithm (uint8_t)
+    contents += struct.pack("B", HashAlgorithm[hash_algo].value)  # HashAlgorithm (uint8_t)
+    contents += struct.pack("I", len(signature))  # Length (uint32_t)
+    contents += struct.pack(f"{len(signature)}s", signature)  # Data (uint8_t[])
+
+    assertPadded(contents, "packV1 contents size is not aligned")
+    return contents
+
+def writeNoteFile(filename, name, sig_algo, hash_algo, signature):
+    name += b"\x00"
+    contents = packV1(sig_algo, hash_algo, signature)
+    with open(filename, "wb") as fd:
+        fd.write(struct.pack("III", len(name), len(contents), 0))
+        fd.write(struct.pack(f"{paddedLength(name)}s", name))
+        fd.write(struct.pack(f"{len(contents)}s", contents))
+
+
+# Validate signing algorithm
+if not hasattr(SignatureAlgorithm, sig_algo):
+    print(f"Unrecognised signing algorithm: {sig_algo}", file=sys.stderr)
+    sys.exit(1)
+
+# Validate hash algorithm
+if not hasattr(HashAlgorithm, hash_algo):
+    print(f"Unrecognised hash algorithm: {hash_algo}", file=sys.stderr)
+    sys.exit(1)
+
+# Read signature from binary file
+with open(signature_binary_file, "rb") as sig_file:
+    signature = sig_file.read()
+
+print("Removing any existing/conflicting section(s)")
+subprocess.call(["objcopy", f"--remove-section={SECTION_NAME}", livepatch])
+
+print("Adding the signature to the livepatch")
+writeNoteFile(SIGNATURE_FILE, bytes(NAME, "UTF-8"), sig_algo, hash_algo, signature)
+subprocess.check_call(["objcopy", "--add-section", f"{SECTION_NAME}={SIGNATURE_FILE}", livepatch])
+
+print(f"Signature written to .note.{NAME}.signature")
diff --git a/sign/sign-file-wrapper b/sign/sign-file-wrapper
deleted file mode 100755
index b006e96a0a09..000000000000
--- a/sign/sign-file-wrapper
+++ /dev/null
@@ -1,6 +0,0 @@
-#!/bin/bash
-
-set -e
-
-./sign-file sha256 signing_key.pem signing_key.x509 "$@"
-chmod +x "$@"
diff --git a/sign/sign-file.c b/sign/sign-file.c
deleted file mode 100644
index 1d3f3a7b6bf1..000000000000
--- a/sign/sign-file.c
+++ /dev/null
@@ -1,323 +0,0 @@
-/* Sign a module file using the given key.
- *
- * Copyright © 2014-2015 Red Hat, Inc. All Rights Reserved.
- * Copyright © 2015      Intel Corporation.
- *
- * Authors: David Howells <dhowells@redhat.com>
- *          David Woodhouse <dwmw2@infradead.org>
- *
- * This program is free software; you can redistribute it and/or
- * modify it under the terms of the GNU Lesser General Public License
- * as published by the Free Software Foundation; either version 2.1
- * of the licence, or (at your option) any later version.
- */
-#define _GNU_SOURCE
-#include <stdio.h>
-#include <stdlib.h>
-#include <stdint.h>
-#include <stdbool.h>
-#include <string.h>
-#include <getopt.h>
-#include <err.h>
-#include <arpa/inet.h>
-#include <openssl/opensslv.h>
-#include <openssl/bio.h>
-#include <openssl/evp.h>
-#include <openssl/pem.h>
-#include <openssl/err.h>
-#include <openssl/engine.h>
-
-/*
- * Use CMS if we have openssl-1.0.0 or newer available - otherwise we have to
- * assume that it's not available and its header file is missing and that we
- * should use PKCS#7 instead.  Switching to the older PKCS#7 format restricts
- * the options we have on specifying the X.509 certificate we want.
- *
- * Further, older versions of OpenSSL don't support manually adding signers to
- * the PKCS#7 message so have to accept that we get a certificate included in
- * the signature message.  Nor do such older versions of OpenSSL support
- * signing with anything other than SHA1 - so we're stuck with that if such is
- * the case.
- */
-#if OPENSSL_VERSION_NUMBER < 0x10000000L
-#define USE_PKCS7
-#endif
-#ifndef USE_PKCS7
-#include <openssl/cms.h>
-#else
-#include <openssl/pkcs7.h>
-#endif
-
-struct module_signature {
-	uint8_t		algo;		/* Public-key crypto algorithm [0] */
-	uint8_t		hash;		/* Digest algorithm [0] */
-	uint8_t		id_type;	/* Key identifier type [PKEY_ID_PKCS7] */
-	uint8_t		signer_len;	/* Length of signer's name [0] */
-	uint8_t		key_id_len;	/* Length of key identifier [0] */
-	uint8_t		__pad[3];
-	uint32_t	sig_len;	/* Length of signature data */
-};
-
-#define PKEY_ID_PKCS7 2
-
-static char magic_number[] = "~Module signature appended~\n";
-
-static __attribute__((noreturn))
-void format(void)
-{
-	fprintf(stderr,
-		"Usage: sign-file [-dp] <hash algo> <key> <x509> <module> [<dest>]\n");
-	exit(2);
-}
-
-static void display_openssl_errors(int l)
-{
-	const char *file;
-	char buf[120];
-	int e, line;
-
-	if (ERR_peek_error() == 0)
-		return;
-	fprintf(stderr, "At main.c:%d:\n", l);
-
-	while ((e = ERR_get_error_line(&file, &line))) {
-		ERR_error_string(e, buf);
-		fprintf(stderr, "- SSL %s: %s:%d\n", buf, file, line);
-	}
-}
-
-static void drain_openssl_errors(void)
-{
-	const char *file;
-	int line;
-
-	if (ERR_peek_error() == 0)
-		return;
-	while (ERR_get_error_line(&file, &line)) {}
-}
-
-#define ERR(cond, fmt, ...)				\
-	do {						\
-		bool __cond = (cond);			\
-		display_openssl_errors(__LINE__);	\
-		if (__cond) {				\
-			err(1, fmt, ## __VA_ARGS__);	\
-		}					\
-	} while(0)
-
-static const char *key_pass;
-
-static int pem_pw_cb(char *buf, int len, int w, void *v)
-{
-	int pwlen;
-
-	if (!key_pass)
-		return -1;
-
-	pwlen = strlen(key_pass);
-	if (pwlen >= len)
-		return -1;
-
-	strcpy(buf, key_pass);
-
-	/* If it's wrong, don't keep trying it. */
-	key_pass = NULL;
-
-	return pwlen;
-}
-
-int main(int argc, char **argv)
-{
-	struct module_signature sig_info = { .id_type = PKEY_ID_PKCS7 };
-	char *hash_algo = NULL;
-	char *private_key_name, *x509_name, *module_name, *dest_name;
-	bool save_sig = false, replace_orig;
-	bool sign_only = false;
-	unsigned char buf[4096];
-	unsigned long module_size, sig_size;
-	unsigned int use_signed_attrs;
-	const EVP_MD *digest_algo;
-	EVP_PKEY *private_key;
-#ifndef USE_PKCS7
-	CMS_ContentInfo *cms;
-	unsigned int use_keyid = 0;
-#else
-	PKCS7 *pkcs7;
-#endif
-	X509 *x509;
-	BIO *b, *bd = NULL, *bm;
-	int opt, n;
-	OpenSSL_add_all_algorithms();
-	ERR_load_crypto_strings();
-	ERR_clear_error();
-
-	key_pass = getenv("KBUILD_SIGN_PIN");
-
-#ifndef USE_PKCS7
-	use_signed_attrs = CMS_NOATTR;
-#else
-	use_signed_attrs = PKCS7_NOATTR;
-#endif
-
-	do {
-		opt = getopt(argc, argv, "dpk");
-		switch (opt) {
-		case 'p': save_sig = true; break;
-		case 'd': sign_only = true; save_sig = true; break;
-#ifndef USE_PKCS7
-		case 'k': use_keyid = CMS_USE_KEYID; break;
-#endif
-		case -1: break;
-		default: format();
-		}
-	} while (opt != -1);
-
-	argc -= optind;
-	argv += optind;
-	if (argc < 4 || argc > 5)
-		format();
-
-	hash_algo = argv[0];
-	private_key_name = argv[1];
-	x509_name = argv[2];
-	module_name = argv[3];
-	if (argc == 5) {
-		dest_name = argv[4];
-		replace_orig = false;
-	} else {
-		ERR(asprintf(&dest_name, "%s.~signed~", module_name) < 0,
-		    "asprintf");
-		replace_orig = true;
-	}
-
-#ifdef USE_PKCS7
-	if (strcmp(hash_algo, "sha1") != 0) {
-		fprintf(stderr, "sign-file: %s only supports SHA1 signing\n",
-			OPENSSL_VERSION_TEXT);
-		exit(3);
-	}
-#endif
-
-	/* Read the private key and the X.509 cert the PKCS#7 message
-	 * will point to.
-	 */
-	if (!strncmp(private_key_name, "pkcs11:", 7)) {
-		ENGINE *e;
-
-		ENGINE_load_builtin_engines();
-		drain_openssl_errors();
-		e = ENGINE_by_id("pkcs11");
-		ERR(!e, "Load PKCS#11 ENGINE");
-		if (ENGINE_init(e))
-			drain_openssl_errors();
-		else
-			ERR(1, "ENGINE_init");
-		if (key_pass)
-			ERR(!ENGINE_ctrl_cmd_string(e, "PIN", key_pass, 0), "Set PKCS#11 PIN");
-		private_key = ENGINE_load_private_key(e, private_key_name, NULL,
-						      NULL);
-		ERR(!private_key, "%s", private_key_name);
-	} else {
-		b = BIO_new_file(private_key_name, "rb");
-		ERR(!b, "%s", private_key_name);
-		private_key = PEM_read_bio_PrivateKey(b, NULL, pem_pw_cb, NULL);
-		ERR(!private_key, "%s", private_key_name);
-		BIO_free(b);
-	}
-
-	b = BIO_new_file(x509_name, "rb");
-	ERR(!b, "%s", x509_name);
-	x509 = d2i_X509_bio(b, NULL); /* Binary encoded X.509 */
-	if (!x509) {
-		ERR(BIO_reset(b) != 1, "%s", x509_name);
-		x509 = PEM_read_bio_X509(b, NULL, NULL, NULL); /* PEM encoded X.509 */
-		if (x509)
-			drain_openssl_errors();
-	}
-	BIO_free(b);
-	ERR(!x509, "%s", x509_name);
-
-	/* Open the destination file now so that we can shovel the module data
-	 * across as we read it.
-	 */
-	if (!sign_only) {
-		bd = BIO_new_file(dest_name, "wb");
-		ERR(!bd, "%s", dest_name);
-	}
-
-	/* Digest the module data. */
-	OpenSSL_add_all_digests();
-	display_openssl_errors(__LINE__);
-	digest_algo = EVP_get_digestbyname(hash_algo);
-	ERR(!digest_algo, "EVP_get_digestbyname");
-
-	bm = BIO_new_file(module_name, "rb");
-	ERR(!bm, "%s", module_name);
-
-#ifndef USE_PKCS7
-	/* Load the signature message from the digest buffer. */
-	cms = CMS_sign(NULL, NULL, NULL, NULL,
-		       CMS_NOCERTS | CMS_PARTIAL | CMS_BINARY | CMS_DETACHED | CMS_STREAM);
-	ERR(!cms, "CMS_sign");
-
-	ERR(!CMS_add1_signer(cms, x509, private_key, digest_algo,
-			     CMS_NOCERTS | CMS_BINARY | CMS_NOSMIMECAP |
-			     use_keyid | use_signed_attrs),
-	    "CMS_add1_signer");
-	ERR(CMS_final(cms, bm, NULL, CMS_NOCERTS | CMS_BINARY) < 0,
-	    "CMS_final");
-
-#else
-	pkcs7 = PKCS7_sign(x509, private_key, NULL, bm,
-			   PKCS7_NOCERTS | PKCS7_BINARY |
-			   PKCS7_DETACHED | use_signed_attrs);
-	ERR(!pkcs7, "PKCS7_sign");
-#endif
-
-	if (save_sig) {
-		char *sig_file_name;
-
-		ERR(asprintf(&sig_file_name, "%s.p7s", module_name) < 0,
-		    "asprintf");
-		b = BIO_new_file(sig_file_name, "wb");
-		ERR(!b, "%s", sig_file_name);
-#ifndef USE_PKCS7
-		ERR(i2d_CMS_bio_stream(b, cms, NULL, 0) < 0,
-		    "%s", sig_file_name);
-#else
-		ERR(i2d_PKCS7_bio(b, pkcs7) < 0,
-			"%s", sig_file_name);
-#endif
-		BIO_free(b);
-	}
-
-	if (sign_only)
-		return 0;
-
-	/* Append the marker and the PKCS#7 message to the destination file */
-	ERR(BIO_reset(bm) < 0, "%s", module_name);
-	while ((n = BIO_read(bm, buf, sizeof(buf))),
-	       n > 0) {
-		ERR(BIO_write(bd, buf, n) < 0, "%s", dest_name);
-	}
-	ERR(n < 0, "%s", module_name);
-	module_size = BIO_number_written(bd);
-
-#ifndef USE_PKCS7
-	ERR(i2d_CMS_bio_stream(bd, cms, NULL, 0) < 0, "%s", dest_name);
-#else
-	ERR(i2d_PKCS7_bio(bd, pkcs7) < 0, "%s", dest_name);
-#endif
-	sig_size = BIO_number_written(bd) - module_size;
-	sig_info.sig_len = htonl(sig_size);
-	ERR(BIO_write(bd, &sig_info, sizeof(sig_info)) < 0, "%s", dest_name);
-	ERR(BIO_write(bd, magic_number, sizeof(magic_number) - 1) < 0, "%s", dest_name);
-
-	ERR(BIO_free(bd) < 0, "%s", dest_name);
-
-	/* Finally, if we're signing in place, replace the original. */
-	if (replace_orig)
-		ERR(rename(dest_name, module_name) < 0, "%s", dest_name);
-
-	return 0;
-}
